WIP: Optimize PK lookup for queries that match exact PK range

bobrik opened the pull request:
Changelog category:

- Performance Improvement

Changelog entry:

Optimize PK lookup for queries that match exact PK range.

### Detailed description / Documentation draft:

Existing code that looks up marks that match the query has a pathological case, when most of the part does in fact match the query. The code works by recursively splitting a part into ranges and then discarding the ranges that definitely do not match the query, based on primary key.

The problem is that it requires visiting every mark that matches the query, making the complexity of this sort of look up O(n).

For queries that match exact range on the primary key, we can find both left and right parts of the range with O(log 2) complexity.

This change implements exactly that.

To engage this optimization, the query must:

* Have a prefix list of the primary key.
* Have only range or single set element constraints for columns.
* Have only AND as a boolean operator.

Consider a table with `(service, timestamp)` as the primary key.

The following conditions will be optimized:

* `service = 'foo'`
* `service = 'foo' and timestamp >= now() - 3600`
* `service in ('foo')`
* `service in ('foo') and timestamp >= now() - 3600 and timestamp <= now`

The following will fall back to previous lookup algorithm:

* `timestamp >= now() - 3600`
* `service in ('foo', 'bar') and timestamp >= now() - 3600`
* `service = 'foo'`

Note that the optimization won't engage when PK has a range expression followed by a point expression, since in that case the range is not continuous.

Trace query logging provides the following messages types of messages, each representing a different kind of PK usage for a part:

```
Used optimized inclusion search over index for part 20200711_5710108_5710108_0 with 9 steps
Used generic exclusion search over index for part 20200711_5710118_5710228_5 with 1495 steps
Not using index on part 20200710_5710473_5710473_0
```

Number of steps translates to computational complexity.

Here's a comparison for before and after for a query over 24h of data:

```
Read 4562944 rows, 148.05 MiB in 45.19249672 sec.,   100966 rows/sec.,   3.28 MiB/sec.
Read 4183040 rows, 135.78 MiB in 0.196279627 sec., 21311636 rows/sec., 691.75 MiB/sec.
```

This is especially useful for queries that read data in order
and terminate early to return "last X things" matching a query.

See #11564 for more thoughts on this.

bobrik committed:
Optimize PK lookup for queries that match exact PK range

Existing code that looks up marks that match the query has a pathological
case, when most of the part does in fact match the query.

The code works by recursively splitting a part into ranges and then discarding
the ranges that definitely do not match the query, based on primary key.

The problem is that it requires visiting every mark that matches the query,
making the complexity of this sort of look up O(n).

For queries that match exact range on the primary key, we can find
both left and right parts of the range with O(log 2) complexity.

This change implements exactly that.

To engage this optimization, the query must:

* Have a prefix list of the primary key.
* Have only range or single set element constraints for columns.
* Have only AND as a boolean operator.

Consider a table with `(service, timestamp)` as the primary key.

The following conditions will be optimized:

* `service = 'foo'`
* `service = 'foo' and timestamp >= now() - 3600`
* `service in ('foo')`
* `service in ('foo') and timestamp >= now() - 3600 and timestamp <= now`

The following will fall back to previous lookup algorithm:

* `timestamp >= now() - 3600`
* `service in ('foo', 'bar') and timestamp >= now() - 3600`
* `service = 'foo'`

Note that the optimization won't engage when PK has a range expression
followed by a point expression, since in that case the range is not continuous.

Trace query logging provides the following messages types of messages,
each representing a different kind of PK usage for a part:

```
Used optimized inclusion search over index for part 20200711_5710108_5710108_0 with 9 steps
Used generic exclusion search over index for part 20200711_5710118_5710228_5 with 1495 steps
Not using index on part 20200710_5710473_5710473_0
```

Number of steps translates to computational complexity.

Here's a comparison for before and after for a query over 24h of data:

```
Read 4562944 rows, 148.05 MiB in 45.19249672 sec.,   100966 rows/sec.,   3.28 MiB/sec.
Read 4183040 rows, 135.78 MiB in 0.196279627 sec., 21311636 rows/sec., 691.75 MiB/sec.
```

This is especially useful for queries that read data in order
and terminate early to return "last X things" matching a query.

See #11564 for more thoughts on this.

CurtizJ reviewed (commented):
Looks ok.
BTW, is it possible to support case, when continuos prefix of primary key is in `WHERE` clause, but not only when there is whole primary key?
Also 2 tests are failing.

bobrik commented:
My first version was to allow a prefix of a PK to engage this. E.g. if PK is `(a, b, c)`, then the following is allowed:

* `(a, b, c)`
* `(a, b)`
* `(a)`

But the following isn't:

* `(a, c)`
* `(b, c)`
* `(b)`
* `(c)`

Is that what you mean? If so, I can totally make it happen.

CurtizJ commented:
Yes, I meant that.

bobrik commented:
I've added support for PK prefix matches and reorganized code a bit to increase code reuse.

Tests should be happy now, the problem was twofold:

* Missing support for monotonic functions — I don't fully understand how to make it work, made it fall back for now.
* Missing pass-by-reference to `may_be_true_in_range()` and final marks, this is fixed.

I've also added some trace logging with amount of matching steps performed, which I intend to add some tests on. Perhaps it can be integrated with `EXPLAIN`, but I think that should be out of scope of this PR.

Let me know what you think.

bobrik committed:
Allow conditions outside of PK with exact range

Conditions that are outside of PK are marked as `unknown` in `KeyCondition`,
so it's safe to allow them, as long as they are always combined by `AND`.

alexey-milovidov commented:
https://github.com/ClickHouse/ClickHouse/issues/13248

canhld94 commented:
Hi @bobrik , sorry for reopening this pr after a long time. This is a nice optimisation, however, I'm curious why it doesn't apply when key condition has monotonic function chain? 
This is quite useful, for example a table that have primary key is `dateTime` and the query with is like `SELECT * FROM TABLE WHEN toYear(key) > 2019 AND toYear(key) < 2021`. In this case, the key condition should be continuous, but the mark range lookup will fallback to exclusive search because of function `toYear`.
From my point of view, I cannot figure out why monotonic functions can affect the continuous of the key range.
Thanks!
